src/core/AudioEngineProfiler.cpp

```diff
@@ -38,10 +41,23 @@ AudioEngineProfiler::AudioEngineProfiler() :
 
 void AudioEngineProfiler::finishPeriod( sample_rate_t sampleRate, fpp_t framesPerPeriod )
 {
-	int periodElapsed = m_periodTimer.elapsed();
+	// Time taken to process all data and fill the audio buffer.
+	const unsigned int periodElapsed = m_periodTimer.elapsed();
+	// Maximum time the processing can take before causing buffer underflow. Convert to us.
+	const uint64_t timeLimit = static_cast<uint64_t>(1000000) * framesPerPeriod / sampleRate;
+
+	// Compute new overall CPU load and apply exponential averaging.
+	// The result is used for overload detection in AudioEngine::CriticalXRuns()
+	// → the weight of a new sample must be high enough to allow relatively fast changes!
+	const int newCpuLoad = 100 * periodElapsed / timeLimit;
+	m_cpuLoad = std::min(newCpuLoad * 0.1f + m_cpuLoad * 0.9f, 100.f);
 
-	const float newCpuLoad = periodElapsed / 10000.0f * sampleRate / framesPerPeriod;
-    m_cpuLoad = std::clamp<int>((newCpuLoad * 0.1f + m_cpuLoad * 0.9f), 0, 100);
+	// Compute detailed load analysis. Can use stronger averaging to get more stable readout.
+	for (int i = 0; i < DetailCount; i++)
+	{
+		const int newLoad = 100 * m_detailTime[i] / timeLimit;
+		m_detailLoad[i].store(std::min(newLoad * 0.05f + m_detailLoad[i] * 0.95f, 100.f), std::memory_order_relaxed);
```

Comment:
The old load could also be loaded with relaxed ordering:
```suggestion
		const auto oldLoad = m_detailLoad[i].load(std::memory_order_relaxed);
		m_detailLoad[i].store(std::min(newLoad * 0.05f + oldLoad * 0.95f, 100.f), std::memory_order_relaxed);
```